gateway/shard/manager.go

```diff
@@ -0,0 +1,314 @@
+			id: gateway.Identifier{
+				IdentifyData:        data,
+				IdentifyShortLimit:  id.IdentifyShortLimit,
+				IdentifyGlobalLimit: id.IdentifyGlobalLimit,
+			},
+		}
+
+		m.shards[i].Shard, err = new(&m, &m.shards[i].id)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to create shard %d/%d", i, len(m.shards)-1)
+		}
+	}
+
+	return &m, nil
+}
+
+// GatewayURL returns the URL to the gateway. The URL will always have the
+// needed gateway parameters appended.
+func (m *Manager) GatewayURL() string {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	return m.gatewayURL
+}
+
+// NumShards returns the total number of shards. Note that, since the number of
+// shards is variable, the caller should not access this while the manager is
+// being rescaled.
+func (m *Manager) NumShards() int {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	return len(m.shards)
+}
+
+// Shard gets the shard with the given ID.
+func (m *Manager) Shard(ix int) Shard {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	if ix >= len(m.shards) {
+		return nil
+	}
+
+	return m.shards[ix]
+}
+
+// FromGuildID returns the Shard and the shard ID for the guild with the given
+// ID.
+func (m *Manager) FromGuildID(guildID discord.GuildID) (shard Shard, ix int) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	ix = int(uint64(guildID>>22) % uint64(len(m.shards)))
+	return m.shards[ix], ix
+}
+
+// ForEach calls the given function on each shard from first to last. To prevent
+// a race condition, the function is given the total number of shards.
+func (m *Manager) ForEach(f func(shard Shard, total int)) {
+	m.mutex.RLock()
+	defer m.mutex.RUnlock()
+
+	for _, g := range m.shards {
+		f(g, len(m.shards))
+	}
+}
+
+// Open opens all gateways handled by this Manager. If an error occurs, Open
+// will attempt to close all previously opened gateways before returning.
+func (m *Manager) Open() error {
+	return m.OpenCtx(context.Background())
+}
+
+// OpenCtx is the context version of Open.
+func (m *Manager) OpenCtx(ctx context.Context) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return openShards(ctx, m.shards)
+}
+
+// Close closes all gateways handled by this Manager; it will stop rescaling if
+// the manager is currently being rescaled. If an error occurs, Close will
+// attempt to close all remaining gateways first, before returning.
+func (m *Manager) Close() error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if m.rescaling != nil {
+		m.rescaling.haltRescale()
+		m.rescaling.rescaleDone.Wait()
+
+		m.rescaling = nil
+	}
+
+	return closeShards(m.shards)
+}
+
+// Rescale rescales the manager asynchronously. The caller MUST NOT call Rescale
+// in the constructor function; doing so WILL cause the state to be inconsistent
+// and eventually crash and burn and destroy us all.
+func (m *Manager) Rescale() {
+	go m.rescale()
+}
+
+func (m *Manager) rescale() {
+	m.mutex.Lock()
+
+	// Exit if we're already rescaling.
+	if m.rescaling != nil {
+		return
+	}
```

Comment:
The mutex won't be unlocked, when returning in L224
```suggestion
	m.mutex.Lock()

	// Exit if we're already rescaling.
	if m.rescaling != nil {
		m.mutex.Unlock()
		return
	}
```